### ISSUE TITLE
docker.registry settings introduced in 0.13.8 affect external Dockerfile's FROM directive

### ISSUE THREAD

**Issue by Gengar003** [rhuss/docker-maven-plugin#378]
Bug Report
===============

Observed Behavior
------------------------------
When a `<docker.registry>` is specified in the docker-maven-plugin's XML configuration, this registry is used to pull base images identified in the `FROM` directive in external Dockerfiles, when those directives do not explicitly specify a registry.

This is inconsistent with a regular "docker build" of the external Dockerfile, which will use dockerhub as the registry when none is specified.

Expected Behavior
------------------------------
When building an image with an external Dockerfile that contains a `FROM` directive with no registry explicitly specified, the base image should be pulled from Dockerhub if no registry is specified.

Workaround
------------------------------
Modify the external Dockerfile by changing

```
FROM centos:7
...
```

to 

```
FROM registry.hub.docker.com/centos:7
...
```

Demonstration
===============

Project Files
--------------------

Consider a project with only these two files in its root:

**pom.xml**
```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
  <modelVersion>4.0.0</modelVersion>
  
  <groupId>org.gengar</groupId>
  <artifactId>override-from</artifactId>
  <version>1.0.0-SNAPSHOT</version>
  <packaging>jar</packaging>
  
  <properties>
    <docker.registry>docker.gengar.org</docker.registry>
    <docker.repo>test</docker.repo>
    <docker.pluginVersion>0.13.8</docker.pluginVersion>
  </properties>
  
  <build>
    <plugins>
      <plugin>
        <groupId>org.jolokia</groupId>
        <artifactId>docker-maven-plugin</artifactId>
        <version>${docker.pluginVersion}</version>
        <configuration>
          <registry>${docker.registry}</registry>
          <images>
            <image>
              <alias>${project.artifactId}-main-image</alias>
              <name>${docker.registry}/${docker.repo}/${project.artifactId}:${project.version}</name>
              <build>
                <assembly>
                  <dockerFileDir>${project.basedir}</dockerFileDir>
                </assembly>
              </build>
            </image>
          </images>
        </configuration>
        <executions>
          <execution>
            <id>build-docker</id>
            <phase>install</phase>
            <goals>
              <goal>build</goal>
            </goals>
          </execution>
        </executions>
      </plugin>
    </plugins>
  </build>
</project>
```

**Dockerfile**
```
FROM centos:7

RUN echo "hello"
```

Maven Build (0.13.8)
-------------------------

I do not have a Docker registry running at docker.gengar.org. Any attempt to pull an image from there will fail. However, this should not be a problem in this example project, because the FROM in the Dockerfile should come from Dockerhub.

```
/tmp/override-from $ mvn docker:build
[INFO] Scanning for projects...
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building override-from 1.0.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- docker-maven-plugin:0.13.8:build (default-cli) @ override-from ---
[ERROR] DOCKER> unable to ping registry endpoint https://docker.gengar.org/v0/
v2 ping attempt failed with error: Get https://docker.gengar.org/v2/: tls: oversized record received with length 20291
 v1 ping attempt failed with error: Get https://docker.gengar.org/v1/_ping: tls: oversized record received with length 20291 (Internal Server Error: 500)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 2.049 s
[INFO] Finished at: 2016-02-08T14:35:12-06:00
[INFO] Final Memory: 16M/309M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.jolokia:docker-maven-plugin:0.13.8:build (default-cli) on project override-from: Unable to pull 'centos:7' from registry 'docker.gengar.org' -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

Maven Build (0.13.7)
-------------------------

In version 0.13.7, the registry-less FROM directive in an external Dockerfile was correctly interpreted to pull from Dockerhub:

```
/tmp/override-from $ mvn docker:build -Ddocker.pluginVersion=0.13.7
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option PermSize=64M; support was removed in 8.0
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=256M; support was removed in 8.0
[INFO] Scanning for projects...
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building override-from 1.0.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- docker-maven-plugin:0.13.7:build (default-cli) @ override-from ---
[INFO] Building tar: /private/tmp/override-from/target/docker/docker.gengar.org/test/override-from/1.0.0-SNAPSHOT/tmp/docker-build.tar
[INFO] DOCKER>   7 Pulling from library/centos
[INFO] DOCKER>   2bf4902415e3 Pulling fs layer
[INFO] DOCKER>   86bcb57631bd Pulling fs layer
[INFO] DOCKER>   c8a648134623 Pulling fs layer
[INFO] DOCKER>   86bcb57631bd Downloading
[INFO] DOCKER>   86bcb57631bd Download complete
[INFO] DOCKER>   c8a648134623 Downloading
[INFO] DOCKER>   c8a648134623 Download complete
[INFO] DOCKER>   2bf4902415e3 Downloading

(snip)

[INFO] DOCKER>   2bf4902415e3 Download complete
[INFO] DOCKER>   Status: Downloaded newer image for centos:7
[INFO] DOCKER> [docker.gengar.org/test/override-from:1.0.0-SNAPSHOT] "override-from-main-image": Built image d0467f786e38
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 37.383 s
[INFO] Finished at: 2016-02-08T14:40:55-06:00
[INFO] Final Memory: 16M/309M
[INFO] ------------------------------------------------------------------------
```

Docker Output
--------------------

And, of course, plain old `docker build` works properly, too:

```
/tmp/override-from $ docker build .
Sending build context to Docker daemon 9.216 kB
Step 1 : FROM centos:7
7: Pulling from library/centos

2bf4902415e3: Pull complete
86bcb57631bd: Pull complete
c8a648134623: Pull complete
Digest: sha256:db4d91ef365f4716517d5cbdb1be54b6d205912038d67bf687a4c2dad9b85928
Status: Downloaded newer image for centos:7
 ---> c8a648134623
Step 2 : RUN echo "hello"
 ---> Running in 03b8c364bd22
hello
 ---> a72d96ece170
Removing intermediate container 03b8c364bd22
Successfully built a72d96ece170
```

### Note ###

If the base image is pulled from Dockerhub manually, or by some other process, this issue will not appear.

**Comment by rhuss** [comment]
There has been a [change](https://github.com/rhuss/docker-maven-plugin/blob/master/doc/changelog.md) added to allow the transparent selection of registries for pulling images in 0.13.8.

You can easily switch this off by using `docker.push.registry` for setting the registry to push to and `docker.pull.registry` for the pull registry. The corresponding configuration properties are `pushRegistry` and `pullRegistry`. Please refer to the [manual](http://ro14nd.de/docker-maven-plugin/registry-handling.html) for details.

Another option is to switch off auto-pulling with `<autoPull>false</autoPull>` in the configuration which won't try to pull a base image if not present.

Does this solve the issue for you ?

**Comment by Gengar003** [comment]
No, this does not solve the issue. I am aware of the changes in 13.8, and run into a different problem there - a problem that I'm not sure is resolvable.

In describing that issue, this issue (#378) will now contain two different issues - however, I believe that the path to fixing both of them is to remove the d-m-p's override of the default registry in external Dockerfiles' FROM directives.

The Issue with registry-handling 0.13.8
==============================

Use-Case
--------------------

I am attempting to build a base-pom that can wrap an arbitrary Dockerfile in the Maven lifecycle. I don't know what kind of FROM Directive may be in the Dockerfiles that use this pom; it may be one of

1. internal, public docker registry (no auth needed)
2. internal, secure docker registry (auth required)
3. internal dockerhub proxy
4. no registry specified

The changes in 13.8 broke my existing configuration, because now the `<docker.registry>` overrides case 4 (no registry specified) and does not pull from dockerhub. 

So, in my attempt to update my pom to work in the 13.8 world, I set `<docker.registry>` to my internal dockerhub proxy. Now, FROM directives in cases 1 through 3 will work (they worked before), but if somebody does not specify a registry in their FROM directive (as any third-party dockerfile from github almost certainly won't), then it will fall back to the internal proxy and still succeed. Plus, we get to use the internal proxy and avoid having to run out through the firewall/vpn and back again.

The Problem
--------------------

However, after pulling from the internal proxy, the d-m-p tags the `docker-proxy.company.internal/image:tag` image as `image:tag` locally (stripping the registry to match the FROM directive, just as you would expect if the `docker` command-line tool had been run).

This works the first time - but on subsequent runs the d-m-p attempts this tagging operation with force=0, and the tagging fails:

After the first successful run, I see these images:

```
REPOSITORY                                            TAG                       IMAGE ID            CREATED             VIRTUAL SIZE
centos                                                7.2.1511                  8d74077f3b19        8 weeks ago         194.6 MB
docker-proxy.company.internal/centos                  7.2.1511                  8d74077f3b19        8 weeks ago         194.6 MB
```

A subsequent run (maven install; hits the `docker:build` goal only) produces this error message:

```
[ERROR] DOCKER> Conflict: Tag centos:7.2.1511 is already set to image 8d74077f3b19, if you want to replace it, please use -f option (Conflict: 409)
```

(inadequate) Workarounds
--------------------

I have verified that I can work around this in two ways:

1. Ban `FROM` directives at my company that do not explicitly specify a registry (require `FROM docker-proxy.company.internal/centos:7` instead of `FROM centos:7`) - or at least accept that users will have to manually, consciously choose to consume the internal proxy because the d-m-p feature that would let me make it automatic, doesn't work.
2. Do not provide a `docker.registry` property in the pom (and thereby be unable to use my internal dockerhub proxy)

And the following would probably also work, though I can't find a configuration option for this: 

1. Tell the docker-maven-plugin to forcibly re-tag images it pulls down (a risky, bad solution)

Conclusion
--------------------

I don't see a safe way for the d-m-p to offer `docker.registry` as "the default registry when none is specified in an external Dockerfile's FROM directive" - and absent that, believe the behavior should be reverted.

**Comment by rhuss** [comment]
To get back the 0.13.7 behaviour to **not** respect any registry setting for pulling images, why not simply omit `docker.registry` (or use `docker.push.registry` only) ? At the moment I can't see how reverting the behaviour to 0.13.7 helps in your use case to provide a default fallback registry for the base images. If you change in your example pom `<registry>` to `<pushRegistry>` you should get exactly the same behaviour as in 0.13.7. If not, please let me know.

That tagging doesn't work is indeed quite unfortunate and should be fixed. IMO overwriting an existing tag shouldn't be that bad since I consider the image name + tag (but without registry part) to be an *unique id* (except when using 'LATEST' which is more or less a snapshot). The registry handling offered by d-m-p works around the fact that Docker itself puts meta information (registry name) into an identifier so that otherwise identically Docker images which are stored in different registries or within a proxy are considered to be different.

An alternative could be to rewrite the Dockerfile's `FROM` to the image name with registry part before the build, so that tagging would not be needed for building the image. 

What do you think about one of this solutions ?

**Comment by Gengar003** [comment]
You are correct that reverting to the 0.13.7 behavior does not allow a fallback/default registry. Absent a fix for the tagging in 0.13.8, 0.13.7's behavior also does not provide a way to "surprise" a developer with *broken* unexpected behavior.

> I consider the image name + tag (but without registry part) to be an unique id (except when using 'LATEST' which is more or less a snapshot)

I am a big proponent of this approach as well. If this is the d-m-p's official stance then great! Let's forcibly tag in 0.13.8; that will solve all of my problems.

However, there may be people out there with a valid (or at least internally-consistent) system in which their non-latest tags are *not* static. In that case, rewriting the registry-less `FROM` directive before the build (as long as the rewrite was hidden inside the d-m-p and didn't actually alter the real file on disk) would solve the problem in a more permissive way.

Either solution would unblock me. I think I lean slightly towards the more-permissive one (rewriting the FROM).
